src/libcore/iter/traits/iterator.rs

```diff
@@ -1855,18 +1855,15 @@ pub trait Iterator {
     /// ```
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn all<F>(&mut self, f: F) -> bool where
+    fn all<F>(&mut self, mut f: F) -> bool where
         Self: Sized, F: FnMut(Self::Item) -> bool
     {
-        #[inline]
-        fn check<T>(mut f: impl FnMut(T) -> bool) -> impl FnMut(T) -> LoopState<(), ()> {
-            move |x| {
-                if f(x) { LoopState::Continue(()) }
-                else { LoopState::Break(()) }
+        while let Some(x) = self.next() {
+            if !f(x) {
+                return false;
```

Comment:
TIL that '`Iterator::all` is that complex.
Here are the result from godbolt: https://godbolt.org/z/lIVoFc

Reply:
Thanks! I'm inexperienced with godbolt, but the results seem to strongly support my simplification, i.e. the microbenchmark static instruction count drops from 29 to 9 and the cycle count drops from 735 to 260. Is that right? Is there anything else in there of note?

Reply:
The MCA results are misleading here, because of the warning at the bottom of the output that "note: program counter updates are ignored" and thus it doesn't understand the loops.  If you consider that the first one is 4x unrolled, the cycle difference is not unreasonable.

Reply:
What is MCA?

Reply:
The code in that godbolt link isn't actually correct:

```rust
while let Some(&x) = a.iter().next() {
```

`a.iter()` is called on every loop iteration so it's only looking at the first element in the slice.

Reply:
Indeed. I updated the link: https://godbolt.org/z/QlHZ0m